modules/lang-mustache/src/main/java/org/elasticsearch/action/search/template/TransportSearchTemplateAction.java

```diff
@@ -0,0 +1,159 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.indices.query.IndicesQueriesRegistry;
+import org.elasticsearch.script.ExecutableScript;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptContext;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.mustache.MustacheScriptEngineService;
+import org.elasticsearch.search.aggregations.AggregatorParsers;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.suggest.Suggesters;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.script.ScriptContext.Standard.SEARCH;
+
+public class TransportSearchTemplateAction extends HandledTransportAction<SearchTemplateRequest, SearchTemplateResponse> {
+
+    private static final String TEMPLATE_LANG = MustacheScriptEngineService.NAME;
+
+    private final ClusterService clusterService;
+    private final ScriptService scriptService;
+    private final TransportSearchAction searchAction;
+    private final IndicesQueriesRegistry queryRegistry;
+    private final AggregatorParsers aggsParsers;
+    private final Suggesters suggesters;
+
+    @Inject
+    public TransportSearchTemplateAction(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                         ActionFilters actionFilters, IndexNameExpressionResolver resolver,
+                                         ClusterService clusterService, ScriptService scriptService,
+                                         TransportSearchAction searchAction, IndicesQueriesRegistry indicesQueryRegistry,
+                                         AggregatorParsers aggregatorParsers, Suggesters suggesters) {
+        super(settings, SearchTemplateAction.NAME, threadPool, transportService, actionFilters, resolver, SearchTemplateRequest::new);
+        this.clusterService = clusterService;
+        this.scriptService = scriptService;
+        this.searchAction = searchAction;
+        this.queryRegistry = indicesQueryRegistry;
+        this.aggsParsers = aggregatorParsers;
+        this.suggesters = suggesters;
+    }
+
+
+    @Override
+    protected void doExecute(SearchTemplateRequest request, ActionListener<SearchTemplateResponse> listener) {
+        new AsyncSearchTemplateAction(request, listener, SEARCH, clusterService.state()).start();
+    }
+
+    class AsyncSearchTemplateAction {
```

Comment:
I think move the logic in this inner class to the body of the `doExecute()` method? (since we do two things: evaluate template, optionally delegate to search api).